File: workUndone/Suite15/OpenFlight_API/samples/scripts/egimageio1.py
##
##
##  Sample file: egimageio1.py
##
##  Objective: Show how to access and manipulate an image.
##
##  Program functions: Reads an image from the command line.
##     Swaps the channels of the image.
##     Writes the swapped image.
##   
##  API functions used:
##     mgInit(), mgExit(),
##     mgReadImage(), mgWriteImage(),
##     mgReadImageAttributes(), mgWriteImageAttributes(),
##
##
##

import sys

# import OpenFlight API module 
from mgapilib import *

def SwapChannels (pixels, newpixels, type, width, height):

	size = width * height

	if type == MIMG_INT: # intensity - 1 channel - no need to swap 
		for i in range (0, size * (type - 1)):
			newpixels[i] = pixels[i]
		return MG_TRUE, newpixels

	elif type == MIMG_INTA:     # intensity alpha - 2 channels 
		iptr1 = 0
		iptr2 = size
		optr1 = 0
		optr2 = size
		i = size
		for i in range (size, 0, -1):
			newpixels[optr1] = pixels[iptr2]
			newpixels[optr2] = pixels[iptr1]
			iptr1 = iptr1 + 1
			iptr2 = iptr2 + 1
			optr1 = optr1 + 1
			optr2 = optr2 + 1
		return MG_TRUE, newpixels

	elif type == MIMG_RGB:     # RGB - 3 channels - only need to swap channels 1 & 3 
		iptr1 = 0
		iptr2 = size
		iptr3 = 2*size
		optr1 = 0
		optr2 = size
		optr3 = 2*size
		for i in range (size, 0, -1):
			newpixels[optr1] = pixels[iptr3]
			newpixels[optr2] = pixels[iptr2]
			newpixels[optr3] = pixels[iptr1]
			iptr1 = iptr1 + 1
			iptr2 = iptr2 + 1
			iptr3 = iptr3 + 1
			optr1 = optr1 + 1
			optr2 = optr2 + 1
			optr3 = optr3 + 1
		return MG_TRUE, newpixels
 
	elif type == MIMG_RGBA:     # RGBA - 4 channels 
		# Note the different way the channels are stored 
		iptr1 = 0
		optr1 = 0
		for i in range (size, 0, -1):
			newpixels[optr1] = pixels[iptr1+3]
			newpixels[optr1+1] = pixels[iptr1+2]
			newpixels[optr1+2] = pixels[iptr1+1]
			newpixels[optr1+3] = pixels[iptr1]
			iptr1 = iptr1 + 4
			optr1 = optr1 + 4
		return MG_TRUE, newpixels
  
	else:
		return MG_FALSE, newpixels
		
	return MG_FALSE, newpixels

def main ():

	# check for correct number of arguments 
	if len(sys.argv) < 3:
		print "\nUsage: ", sys.argv[0], " <input_texture_filename> <output_texture_filename>\n"
		return

   # initialize the OpenFlight API
   # always call mgInit BEFORE any other OpenFlight API calls 
   #
	mgInit (None, None)

	# Read the input image, pixels is allocated 
	status, pixels, type, width, height = mgReadImage (sys.argv[1])

	if status:
		print "Error reading image: ",sys.argv[1], "status:", status, "\n"
		return

	# Read the input image attributes 
	imgAttrRec = mgReadImageAttributes (sys.argv[1])

	if not imgAttrRec:
		print "Error reading image attributes for: ",sys.argv[1], "\n"
		return

	newpixels = mgunsignedchar(width * height * (type - 1))
	if not newpixels:
		print "Error allocating memory for swapped image\n"
		return

	# Swap the channels 
	if not SwapChannels (pixels, newpixels, type, width, height):
		print "Error swapping the channels\n"
		return

	# Write the swapped image 
	status = mgWriteImage (sys.argv[2], newpixels, type, width, height, MG_FALSE)
	if status:
		print "Error writing image: ",sys.argv[2], "\n"
		return

	# Write the swapped image attributes 
	if not mgWriteImageAttributes (sys.argv[2], imgAttrRec):
		print "Error writing image attributes for: ",sys.argv[2], "\n"
		return

	# always call mgExit() AFTER all OpenFlight API calls 
	mgExit ()

main()

File: workUndone/Suite15/OpenFlight_API/samples/apps/src/egimageio1.c
/*******************************************************************************

  Sample file: EGIMAGEIO1.C

  Objective: Show how to access and manipulate an image.

  Program functions: Reads an image from the command line.
     Swaps the channels of the image.
     Writes the swapped image.
   
  API functions used:
     mgInit(), mgExit(),
     mgReadImage(), mgWriteImage(),
     mgReadImageAttributes(), mgWriteImageAttributes(),
	  mgMalloc(), mgFree()

 ******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* include all OpenFlight API headers */
#include <mgapiall.h>

static int SwapChannels (unsigned char* pixels, unsigned char* newpixels,
									int type, int width, int height)
{
   unsigned char *iptr1, *iptr2, *iptr3;
   unsigned char *optr1, *optr2, *optr3;
   int i, size;

   size = width * height;
   switch (type)
   {
   case MIMG_INT:     /* intensity - 1 channel - no need to swap */
      memcpy (newpixels, pixels, size * (type -1));
      return (MG_TRUE);
		break;
   
	case MIMG_INTA:     /* intensity alpha - 2 channels */
      iptr1 = pixels;
      iptr2 = pixels + size;
      optr1 = newpixels;
      optr2 = newpixels + size;
      i = size;
      while (i--)
      {
         *optr1++ = *iptr2++;
         *optr2++ = *iptr1++;
      }
      return (MG_TRUE);
		break;
   
	case MIMG_RGB:     /* RGB - 3 channesl - only need to swap channels 1 & 3 */
      iptr1 = pixels;
      iptr2 = pixels + size;
      iptr3 = pixels + 2*size;
      optr1 = newpixels;
      optr2 = newpixels + size;
      optr3 = newpixels + 2*size;
      i = size;
      while (i--)
      {
         *optr1++ = *iptr3++;
         *optr2++ = *iptr2++;
         *optr3++ = *iptr1++;
      }
      return (MG_TRUE);
		break;
 
	case MIMG_RGBA:     /* RGBA - 4 channels */
		/* Note the different way the channels are stored */
      iptr1 = pixels;
      optr1 = newpixels;
      i = size;
      while (i--)
      {
         *optr1 = *(iptr1+3);
         *(optr1+1) = *(iptr1+2);
         *(optr1+2) = *(iptr1+1);
         *(optr1+3) = *iptr1;
         iptr1 += 4;
         optr1 += 4;
      }
      return (MG_TRUE);
		break;
  
	default:
      return (MG_FALSE);
		break;
   }
	return (MG_FALSE);
}

int main (int argc, char* argv[])
{
   int type, width, height, status;
   unsigned char *pixels, *newpixels;
   mgrec* imgAttrRec;

   /* check for proper arguments */
   if (argc < 3) {
      printf ("\nUsage: %s <input_texture_filename> <output_texture_filename>\n", argv[0]);
      exit (0);
   }

   /* initialize the OpenFlight API
      always call mgInit BEFORE any other OpenFlight API calls 
      */
   mgInit (&argc, argv);

   /* Read the input image, pixels is allocated, remember to mgFree */
   status = mgReadImage (argv[1], &pixels, &type, &width, &height);

   if (status != 0) {
		printf ("Error reading image: %s\n", argv[1]);
      exit (EXIT_FAILURE);
   }

   /* Read the input image attributes */
   imgAttrRec = mgReadImageAttributes (argv[1]);

   if (imgAttrRec == NULL) {
      printf ("Error reading image attributes for: %s\n", argv[1]);
      exit (EXIT_FAILURE);     
   }

   /* Allocate the swapped image */
   newpixels = (unsigned char*) mgMalloc (width * height * (type -1));
   if (!newpixels) {
      printf ("Error allocating memory for swapped image\n");
      exit (EXIT_FAILURE);     
   }

   /* Swap the channels */
   if (!SwapChannels (pixels, newpixels, type, width, height)) {
      printf ("Error swapping the channels\n");
      exit (EXIT_FAILURE);     
   }

   /* Write the swapped image */
   status = mgWriteImage (argv[2], newpixels, type, width, height, MG_FALSE);
   if (status != 0) {
		printf ("Error writing image: %s\n", argv[2]);
      exit (EXIT_FAILURE);
   }

   /* Write the swapped image attributes */
   if (!mgWriteImageAttributes (argv[2], imgAttrRec)) {
      printf ("Error writing image attributes for: %s\n", argv[2]);
      exit (EXIT_FAILURE);     
   }

   mgFree (pixels);
   mgFree (newpixels);

	/* always call mgExit() AFTER all OpenFlight API calls */
	mgExit ();

   exit (0);
}
